### PR TITLE
Ride BigInt support

### PR DESCRIPTION
**Pull request by alexeykiselev**

### CODE REVIEW

File: pkg/ride/math/math.go [hunk 1]
```diff
@@ -12,7 +14,15 @@ var (
 	ten  = decimal.New(10, 0)
 )
 
-func convertToResult(v *decimal.Big, scale int, mode decimal.RoundingMode) (int64, error) {
+func checkScales(bs, es, rs int) bool {
```
Comment (esuwu) [inline pkg/ride/math/math.go:17]: leave here a comment about these "bs", "es", "rs" please

File: pkg/ride/math/math.go [hunk 2]
```diff
@@ -64,52 +99,98 @@ func Fraction(value, numerator, denominator int64) (int64, error) {
...
+func FloorDivBigInt(x, y *big.Int) *big.Int {
+	var r *big.Int
+	if x.Sign() == y.Sign() {
+		if x.Cmp(y) < 0 {
+			// abs(y-x)/2 + x
+			r = y.Sub(y, x)
+			r = r.Abs(r)
+			r = r.Div(r, big.NewInt(2))
+			r = r.Add(r, x)
+		} else {
```
Comment (esuwu) [inline pkg/ride/math/math.go:176]: we don't need `else` here. You should make `return` just before this line in the end of `if`

File: pkg/ride/math/math.go [hunk 3]
```diff
@@ -64,52 +99,98 @@ func Fraction(value, numerator, denominator int64) (int64, error) {
...
+			// abs(x-y)/2 + y
+			r = x.Sub(x, y)
+			r = r.Abs(r)
+			r = r.Div(r, big.NewInt(2))
+			r = r.Add(r, y)
+		}
```
Comment (esuwu) [inline pkg/ride/math/math.go:182]: `return`

File: pkg/ride/math/math.go [hunk 4]
```diff
@@ -64,52 +99,98 @@ func Fraction(value, numerator, denominator int64) (int64, error) {
...
+	} else {
```
Comment (esuwu) [inline pkg/ride/math/math.go:183]: without `else`

File: pkg/ride/math/math.go [hunk 5]
```diff
@@ -64,52 +99,98 @@ func Fraction(value, numerator, denominator int64) (int64, error) {
...
+		d := x.Add(x, y)
+		two := big.NewInt(2)
+		zero := big.NewInt(0)
+		d2 := big.NewInt(0).Mod(d, two)
+		if d.Cmp(zero) >= 0 || d2.Cmp(zero) == 0 {
+			r = d.Div(d, two)
```
Comment (esuwu) [inline pkg/ride/math/math.go:189]: `return`

File: pkg/ride/math/math.go [hunk 6]
```diff
@@ -64,52 +99,98 @@ func Fraction(value, numerator, denominator int64) (int64, error) {
...
+		} else {
```
Comment (esuwu) [inline pkg/ride/math/math.go:190]: no `else` here

File: pkg/ride/math/math.go [hunk 7]
```diff
@@ -27,31 +37,56 @@ func convertToResult(v *decimal.Big, scale int, mode decimal.RoundingMode) (int6
...
 }
 
+func convertToBigIntResult(v *decimal.Big, scale int, mode decimal.RoundingMode) (*big.Int, error) {
+	if v.IsNaN(0) || v.IsInf(0) {
+		return nil, errors.New("result is NaN or Infinity")
+	}
+	context := decimal.Context128
+	context.RoundingMode = mode
+	r := decimal.WithContext(context).Set(v)
+	s := decimal.WithContext(decimal.Context128).SetMantScale(int64(scale), 0)
```
Comment (esuwu) [inline pkg/ride/math/math.go:47]: put a comment here about `r`, `s` and `m` please

**Review by esuwu — CHANGES_REQUESTED** [review]

**Review by esuwu — CHANGES_REQUESTED** [review]

File: pkg/ride/functions_bigint.go [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+		return rideBigInt{}, errors.Errorf("argument 1 is not of type 'BigInt' but '%s'", args[0].instanceOf())
+	}
+	return l, nil
+}
+
+func twoBigIntArgs(args []rideType) (rideBigInt, rideBigInt, error) {
+	if len(args) != 2 {
+		return rideBigInt{}, rideBigInt{}, errors.Errorf("%d is invalid number of arguments, expected 2", len(args))
+	}
+	if args[0] == nil {
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/ride/functions_bigint.go:49]: Such function already exists and used, but this copy-paste is intended. Won't fix.

File: pkg/ride/functions_bigint.go [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+	e := big.NewInt(0).Set(exponent.v)
+	r, err := math.LogBigInt(b, e, int(bp), int(ep), int(rp), round)
+	if err != nil {
+		return nil, errors.Wrap(err, "logBigInt")
+	}
+	return rideBigInt{v: r}, nil
+}
+
+func toBigInt(_ Environment, args ...rideType) (rideType, error) {
+	i, err := intArg(args)
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/ride/functions_bigint.go:172]: I don't get you, it's already like this.

File: pkg/ride/functions_bigint.go [hunk 10]
```diff
@@ -0,0 +1,658 @@
...
+		return nil, errors.Errorf("bytesToBigIntLim: offset %d is out of range [0; %d]", offset, len(bts)-1)
+	}
+	size, ok := args[2].(rideInt)
+	if !ok {
+		return nil, errors.Errorf("bytesToBigIntLim: argument 3 is not of type 'Int' but '%s'", args[2].instanceOf())
+	}
+	if size <= 0 || size > 64 {
+		return nil, errors.Errorf("bytesToBigIntLim: size %d is out of ranger [1; 64]", size)
+	}
+	end := int(offset + size)
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/ride/functions_bigint.go:491]: No, it's not about precision, it's about slicing bytes in rideBytes.

File: pkg/ride/functions_bigint.go [hunk 11]
```diff
@@ -0,0 +1,658 @@
...
+		r := math.FloorDivBigInt(x.Add(x, y), big.NewInt(2))
+		return rideBigInt{v: r}, nil
+	}
+}
+
+func minMaxBigInt(items []*big.Int) (*big.Int, *big.Int) {
+	if len(items) == 0 {
+		panic("empty slice")
+	}
+	max := items[0]
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/ride/functions_bigint.go:578]: Yep

File: pkg/ride/functions_bigint.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/ride/functions_bigint.go:170]: sorry, missed it

File: pkg/ride/functions_bigint.go [hunk 12]
```diff
@@ -0,0 +1,655 @@
...
+	}
+	d := big.NewInt(0).Set(v3.v)
+	if d.Cmp(zeroBigInt) == 0 {
+		return nil, errors.New("fractionBigIntRounds: division by zero")
+	}
+	round, err := roundingMode(args[3])
+	if err != nil {
+		return nil, errors.Wrap(err, "fractionBigIntRounds")
+	}
+	p := v.Mul(v, n)
```
Comment (esuwu) [inline pkg/ride/functions_bigint.go:302]: leave a comment here

File: pkg/ride/functions_bigint.go [hunk 13]
```diff
@@ -0,0 +1,655 @@
...
+	}
+	items, err := toBigIntSlice(list)
+	if err != nil {
+		return nil, errors.Wrap(err, "medianListBigInt")
+	}
+	sort.Sort(items)
+	half := size / 2
+	if size%2 == 1 {
+		return rideBigInt{v: items[half]}, nil
+	} else {
```
Comment (esuwu) [inline pkg/ride/functions_bigint.go:564]: don't need `else` here

File: pkg/ride/math/math.go [hunk 14]
```diff
@@ -12,7 +14,15 @@ var (
 	ten  = decimal.New(10, 0)
 )
 
-func convertToResult(v *decimal.Big, scale int, mode decimal.RoundingMode) (int64, error) {
+func checkScales(baseScale, exponentScale, resultScale int) bool {
+	return baseScale >= 0 && baseScale <= 8 && exponentScale >= 0 && exponentScale <= 8 && resultScale >= 0 && resultScale <= 8
+}
+
+func checkScalesBigInt(baseScale, exponentScale, resultScale int) bool {
+	return baseScale >= 0 && baseScale <= 18 && exponentScale >= 0 && exponentScale <= 18 && resultScale >= 0 && resultScale <= 18
```
Comment (esuwu) [inline pkg/ride/math/math.go:22]: magic constants

**Review by esuwu — CHANGES_REQUESTED** [review]

File: pkg/ride/functions_bigint.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/ride/functions_bigint.go:47]: then why wouldn't you use it?

**Comment by esuwu** [comment]
still need variables with the right name especially for those ones `18`, `8`, `511`,  maybe for `0x80` as well

**Review by esuwu — APPROVED** [review]

**Pull request merged**